src/k8s/requests/v2v/importVmware.js

```diff
@@ -0,0 +1,263 @@
+  const password =
+    get(getSetting(PROVIDER_VMWARE_USER_PWD_AND_CHECK_KEY), PROVIDER_VMWARE_USER_PWD_KEY) ||
+    extractFromSecret('password');
+
+  const hostname = getSetting(PROVIDER_VMWARE_HOSTNAME_KEY) || extractFromSecret('url');
+
+  const sourceDisks = storages
+    .filter(storage => storage.storageType === STORAGE_TYPE_EXTERNAL_IMPORT)
+    .map(({ data }) => data.fileName);
+
+  const conversionData = {
+    daemonize: false,
+
+    vm_name: getSetting(PROVIDER_VMWARE_VM_KEY),
+    transport_method: 'vddk',
+
+    vmware_fingerprint: vmWareData && vmWareData.thumbPrint,
+    vmware_uri: `vpx://${username}@${hostname}${vmWareData && vmWareData.hostPath}?no_verify=1`,
+    vmware_password: password,
+
+    source_disks: sourceDisks,
+  };
+
+  const conversionPodSecret = await k8sCreate(
+    SecretModel,
+    buildConversionPodSecret({ namespace, data: conversionData })
+  );
+
+  return {
+    resultObjects: [...resultObjects, conversionPodSecret],
+    conversionPodSecret,
+  };
+};
+
+// // Start of the Conversion pod is blocked until the PVCs are bound
+const startConversionPod = async (
+  getSetting,
+  networks,
+  storages,
+  { k8sCreate, k8sPatch },
+  { resultObjects, serviceAccount, role, roleBinding, mappedStorages, conversionPodSecret }
+) => {
+  const namespace = getSetting(NAMESPACE_KEY);
+  const volumes = [];
+  const volumeMounts = [];
+
+  mappedStorages
+    .filter(({ storageType }) =>
+      [STORAGE_TYPE_EXTERNAL_IMPORT, STORAGE_TYPE_EXTERNAL_V2V_TEMP, STORAGE_TYPE_EXTERNAL_V2V_VDDK].includes(
+        storageType
+      )
+    )
+    .forEach(storage => {
+      volumeMounts.push(asVolumenMount(storage));
+      volumes.push(asVolume(storage));
+    });
+
+  const conversionPod = await k8sCreate(
+    PodModel,
+    buildConversionPod({
+      volumes,
+      volumeMounts,
+      namespace,
+      serviceAccountName: getName(serviceAccount),
+      secretName: getName(conversionPodSecret),
+    })
+  );
+
+  if (conversionPod) {
+    const newOwnerReferences = [buildOwnerReference(conversionPod)];
+    const patchPvcs = mappedStorages
+      .filter(storage => storage.storageType === STORAGE_TYPE_EXTERNAL_V2V_TEMP)
+      .map(({ data }) => data.pvc);
+
+    await [serviceAccount, role, roleBinding, conversionPodSecret, ...patchPvcs].forEach(async object => {
+      const patches = [buildAddOwnerReferencesPatch(object, newOwnerReferences)];
+      await k8sPatch(findModel(object), object, patches);
+      resultObjects = replaceUpdatedObject(resultObjects, object);
+    });
+  }
+
+  return {
+    resultObjects: [...resultObjects, conversionPod],
+    conversionPod,
+  };
+};
+
+export const importVmwareVm = async (getSetting, networks, storages, { k8sCreate, k8sPatch }) =>
+  [
+    createProviderSecret,
+    createConversionPodSecret,
+    resolveRolesAndServiceAccount,
+    resolveStorages,
+    startConversionPod,
+  ].reduce(
```

Comment:
Vivat simplicity! Wouldn't it be more readable for future generations to avoid such constructs?

Reply:
I think it is quite readable. And it gives a clear contract for each step what data to expect. I have one refactoring lined up, which will simplify this use case a bit (error rollback).